Travis: Add cpu trap flag tests

andrewbird opened the pull request:
Run as `python3 test/test_dos.py PPDOSGITTestCase.test_cpu_trap_flag`

1. Can you review my changes to your test case?
1. Can you check this fails on your system with current devel (should fail due to fixup message)?
2. Then if you roll back devel to before the 'fixup' commit, check that it still fails. (should fail due to incorrect value)?

Thanks!

andrewbird commented:
I pushed a new version that checks a list for known bad cpus and reports 'unexpected success' or 'unknown bad cpu' along with cpu info. At the moment I have only my machine listed, so I get a 'unexpected success'. If you run it you should get 'unknown bad cpu'. Can you report the cpu string it returns so that I can add it to the known bad list, please? Note that I split the test so you now have to run as:
`python test/test_dos.py PPDOSGITTestCase.test_cpu_trap_flag_emulated PPDOSGITTestCase.test_cpu_trap_flag_kvm`

andrewbird commented:
And you'll need to add the 'python3-cpuinfo' package to your machine.

stsp commented on test/func_cpu_trap_flag.py:
```diff
@@ -0,0 +1,88 @@
+    pushw   $0x3003
+    pushw   $0x3303
+    popf
+    popf
```
Please change both to popfw.
Its the same thing, but for the test-case
its better to be explicit, especially since
popfl actually "kinda works".

andrewbird replied:
okay

stsp commented on test/func_cpu_trap_flag.py:
```diff
@@ -0,0 +1,88 @@
+exit:
+    movw    $0x4c00, %ax
+    int     $0x21
+    ret
```
Why not just ret?
Or do you mind the retcode?

andrewbird replied:
I don't really care about the retcode whilst in the suite, but if run standalone someone might.

stsp commented:
> Can you review my changes to your test case?

Done.

> Can you check this fails on your system with current devel (should fail due to fixup message)?

```
ERROR: Test PP-DOS-GIT CPU Trap Flag KVM                                               
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/stas/src/dosemu2-devel/test/test_dos.py", line 5313, in test_cpu_trap_flag_kvm
    cpu_trap_flag(self, 'kvm')
  File "/home/stas/src/dosemu2-devel/test/func_cpu_trap_flag.py", line 81, in cpu_trap_flag
    if cpu['brand'] in knownbad:
KeyError: 'brand'
```

Seems like an odd error?

> Then if you roll back devel to before the 'fixup' commit, check that it still fails. (should fail due to incorrect value)?

Not by a roll-back, as it will also remove the
test. But a revert should do, so:
```
Result is (4)


C:\>rem end


======================================================================
FAIL: Test PP-DOS-GIT CPU Trap Flag KVM                                               
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/stas/src/dosemu2-devel/test/test_dos.py", line 5313, in test_cpu_trap_flag_kvm
    cpu_trap_flag(self, 'kvm')
  File "/home/stas/src/dosemu2-devel/test/func_cpu_trap_flag.py", line 63, in cpu_trap_flag
    self.assertIn("Result is (1)", results)
AssertionError: 'Result is (1)' not found in 'testit.bat\r\r\n\r\r\nC:\\>c:\\cputrapf\r\r\nResult is (4)\r\n\r\r\nC:\\>'
```

Is it good??

stsp commented:
Is there any better way of reporting the
test failures, rather than with an assertion
and a trace-back?

stsp commented:
> Can you report the cpu string it returns

I don't see it returning any.
But I copy/pasted all I have.

andrewbird commented:
Did you see that I pushed some changes since yesterday?

andrewbird commented:
> Is there any better way of reporting the
test failures, rather than with an assertion
and a trace-back?

Not that I know of. I guess this is a result of using a python test framework where tracebacks are actually useful.

stsp commented:
> Did you see that I pushed some changes since yesterday?

No but I did the pull today, so everything is up to date.

andrewbird commented:
Please can you paste the results of 
`python3 -c "import cpuinfo ; print(cpuinfo.get_cpu_info())"`

stsp commented:
```
{'python_version': '3.9.1.candidate.1 (64 bit)', 'cpuinfo_version': [7, 0, 0], 'cpuinfo_version_string': '7.0.0', 'arch': 'X86_64', 'bits': 64, 'count': 8, 'arch_string_raw': 'x86_64', 'vendor_id_raw': 'AuthenticAMD', 'brand_raw': 'AMD FX(tm)-8350 Eight-Core Processor', 'hz_advertised_friendly': '3.7604 GHz', 'hz_actual_friendly': '3.7604 GHz', 'hz_advertised': [3760401000, 0], 'hz_actual': [3760401000, 0], 'model': 2, 'family': 21, 'flags': ['3dnowprefetch', 'abm', 'aes', 'aperfmperf', 'apic', 'arat', 'avx', 'bmi1', 'clflush', 'cmov', 'cmp_legacy', 'constant_tsc', 'cpb', 'cpuid', 'cr8_legacy', 'cx16', 'cx8', 'de', 'decodeassists', 'extapic', 'extd_apicid', 'f16c', 'flushbyasid', 'fma', 'fma4', 'fpu', 'fxsr', 'fxsr_opt', 'ht', 'hw_pstate', 'ibpb', 'ibs', 'lahf_lm', 'lbrv', 'lm', 'mca', 'mce', 'misalignsse', 'mmx', 'mmxext', 'monitor', 'msr', 'mtrr', 'nodeid_msr', 'nonstop_tsc', 'nopl', 'npt', 'nrip_save', 'nx', 'osvw', 'pae', 'pat', 'pausefilter', 'pclmulqdq', 'pdpe1gb', 'perfctr_core', 'perfctr_nb', 'pfthreshold', 'pge', 'pni', 'popcnt', 'pse', 'pse36', 'rdtscp', 'rep_good', 'sep', 'skinit', 'ssbd', 'sse', 'sse2', 'sse4_1', 'sse4_2', 'sse4a', 'ssse3', 'svm', 'svm_lock', 'syscall', 'tbm', 'tce', 'topoext', 'tsc', 'tsc_scale', 'vmcb_clean', 'vme', 'vmmcall', 'wdt', 'xop', 'xsave'], 'l3_cache_size': 2097152, 'l2_cache_size': '8 MiB', 'l1_data_cache_size': '64 KiB', 'l1_instruction_cache_size': '256 KiB'}
```

andrewbird commented:
Thanks. It seems the 'brand' key got renamed to 'brand_raw', so we'll use the first one that exists. I added your CPU to the known bad ones and removed mine. So you should now get a skip and a pass, whereas I get two passes.

stsp commented:
```
Test PP-DOS-GIT CPU Trap Flag emulated                                           ... ok (  1.94s)
Test PP-DOS-GIT CPU Trap Flag KVM                                                ... skipped "known bad cpu 'AMD FX(tm)-8350 Eight-Core Processor'"
```

stsp commented:
> Will skip if fixup is used, will fail if somehow
>    it doesn't need to use the fixup.

How is so?
If you skip something, you can't find it
didn't need a fixup.

andrewbird commented:
> How is so?

The test is always run and only the reporting is different.
  1. If the result is anything other than '1' it's reported as fail.
  1. If we detected that the fixup was used and we were not expecting it to be, it's reported as 'bad cpu' failure.
  1. If we detected that the fixup was used and we were expecting it to be, it's reported as a 'known bad cpu' skip.
  1. If we detected that the fixup wasn't used and we were not expecting it to be, it's reported as pass
  1. If we detected that the fixup wasn't used and we were expecting it to be, it's reported as 'unexpected success' failure

I could report 3. as a pass, but there's no opportunity to print any additional message with it. As you've noticed 'skip' is bad terminology, but I wanted to indicate that the workaround is in play.

andrewbird commented:
Oh one more thing. I'm testing this with 'kvm' and 'emulated', should I really be testing two types of emulation 'jit' and 'sim' alongside 'kvm'?

stsp commented:
At least for the sake of that test, there
is no compeling need to run it in a simulator,
as it seems to detect the bug in a real CPU.
The rest depends on whether you want to
run it along with all generic tests.

andrewbird commented:
I found a way to extend '3.' above to PASS  with message, is it better now (obviously I faked the test here)?
~~~
Test PP-DOS-GIT CPU Trap Flag KVM                                                ... ok (  3.45s) known bad cpu 'Intel(R) Core(TM)2 Quad CPU    Q6600  @ 2.40GHz'
~~~

stsp commented:
Yes, looks much better.

andrewbird commented:
I updated it with regard to the message on pass, if you could just check with the following, it should be good to merge.
`python test/test_dos.py PPDOSGITTestCase.test_cpu_trap_flag_emulated PPDOSGITTestCase.test_cpu_trap_flag_kvm`

stsp commented:
```
Test PP-DOS-GIT CPU Trap Flag emulated                                           ... ok (  1.33s)
Test PP-DOS-GIT CPU Trap Flag KVM                                                ... ok (  1.54s) known bad cpu 'AMD FX(tm)-8350 Eight-Core Processor'
```

andrewbird commented:
Thanks, that looks fine then.

stsp commented:
Thanks!

stsp commented:
These tests are cool.
I have very few bisects lately.
Would be nice to integrate @jschwartzenberg 's
windows installer test.
Does it still work?

andrewbird commented:
Thanks, I'm glad they are being of some value! About the windows installer test, I don't know it myself, but if we add it I should probably split it and the other less DOS version dependant tests from the main test run, else we'll be eating too many travis minutes for little gain.

stsp commented:
> Thanks, I'm glad they are being of some value!

Well, they only annoy me with those
"devel is broken" e-mails. :) So its quite
difficult to compare retroactively, but I
think the amount of bisects is reduced
because of them, and bisects are a PITA
with fdpp, as you know.

Yeah, windows test should be in a cron
weekly job. It does crazy things like
screen-shotting and OCRing windows
screen to find out the needed buttons
and messages... But I think he actually
got it all to work in the past.
